# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/CanonicalBrowsingContext.h
# Commit: d6f212c67002
# Full Hash: d6f212c670027f013f8453181d9b1cfc9b4816dd
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2021-03-31 21:54:44
# Regressor Bug: 1699721
# File Overlap Count: 1
# Description:
#   Bug 1699721 - Part 2: Track BrowserParent lifecycles during process switches, r=kmag
#   
#   This patch contains a large number of changes around the process switching
#   mechanism in order to avoid issues which are caused by a mismatched
#   understanding of the state of the process switch between processes in the
# ==============================================================================

diff -r ef06d9764cf1 -r d6f212c67002 docshell/base/CanonicalBrowsingContext.h
--- a/docshell/base/CanonicalBrowsingContext.h	Wed Mar 31 15:37:48 2021 +0000
+++ b/docshell/base/CanonicalBrowsingContext.h	Wed Mar 31 15:37:49 2021 +0000
@@ -39,6 +39,7 @@
 namespace dom {
 
 class BrowserParent;
+class BrowserBridgeParent;
 class FeaturePolicy;
 struct LoadURIOptions;
 class MediaController;
@@ -84,10 +85,6 @@
 
   void SetOwnerProcessId(uint64_t aProcessId);
 
-  void SetInFlightProcessId(uint64_t aProcessId);
-  void ClearInFlightProcessId(uint64_t aProcessId);
-  uint64_t GetInFlightProcessId() const { return mInFlightProcessId; }
-
   // The ID of the BrowsingContext which caused this BrowsingContext to be
   // opened, or `0` if this is unknown.
   // Only set for toplevel content BrowsingContexts, and may be from a different
@@ -215,6 +212,7 @@
   void SetCurrentRemoteURI(nsIURI* aCurrentRemoteURI);
 
   BrowserParent* GetBrowserParent() const;
+  void SetCurrentBrowserParent(BrowserParent* aBrowserParent);
 
   // Internal method to change which process a BrowsingContext is being loaded
   // in. The returned promise will resolve when the process switch is completed.
@@ -287,6 +285,14 @@
   void SetRestoreData(SessionStoreRestoreData* aData);
   void RequestRestoreTabContent(WindowGlobalParent* aWindow);
 
+  // Called when a BrowserParent for this BrowsingContext has been fully
+  // destroyed (i.e. `ActorDestroy` was called).
+  void BrowserParentDestroyed(BrowserParent* aBrowserParent,
+                              bool aAbnormalShutdown);
+
+  void StartUnloadingHost(uint64_t aChildID);
+  void ClearUnloadingHost(uint64_t aChildID);
+
  protected:
   // Called when the browsing context is being discarded.
   void CanonicalDiscard();
@@ -319,10 +325,14 @@
     friend class CanonicalBrowsingContext;
 
     ~PendingRemotenessChange();
+    void ProcessLaunched();
     void ProcessReady();
     void Finish();
     void Clear();
 
+    nsresult FinishTopContent();
+    nsresult FinishSubframe();
+
     RefPtr<CanonicalBrowsingContext> mTarget;
     RefPtr<RemotenessPromise::Private> mPromise;
     RefPtr<GenericPromise> mPrepareToChangePromise;
@@ -350,6 +360,16 @@
       const nsID& aChangeID,
       const CallerWillNotifyHistoryIndexAndLengthChanges& aProofOfCaller);
 
+  struct UnloadingHost {
+    uint64_t mChildID;
+    nsTArray<std::function<void()>> mCallbacks;
+  };
+  nsTArray<UnloadingHost>::iterator FindUnloadingHost(uint64_t aChildID);
+
+  // Called when we want to show the subframe crashed UI as our previous browser
+  // has become unloaded for one reason or another.
+  void ShowSubframeCrashedUI(BrowserBridgeParent* aBridge);
+
   // XXX(farre): Store a ContentParent pointer here rather than mProcessId?
   // Indicates which process owns the docshell.
   uint64_t mProcessId;
@@ -357,10 +377,6 @@
   // Indicates which process owns the embedder element.
   uint64_t mEmbedderProcessId;
 
-  // The ID of the former owner process during an ownership change, which may
-  // have in-flight messages that assume it is still the owner.
-  uint64_t mInFlightProcessId = 0;
-
   uint64_t mCrossGroupOpenerId = 0;
 
   // This function will make the top window context reset its
@@ -369,6 +385,10 @@
   // entries are added or replaced.
   void ResetSHEntryHasUserInteractionCache();
 
+  RefPtr<BrowserParent> mCurrentBrowserParent;
+
+  nsTArray<UnloadingHost> mUnloadingHosts;
+
   // The current URI loaded in this BrowsingContext. This value is only set for
   // BrowsingContexts loaded in content processes.
   nsCOMPtr<nsIURI> mCurrentRemoteURI;